src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -203,42 +211,148 @@ MakeIntermediateResultHTAB()
 
 /*
  * FindAllWorkerNodesUsingSubplan creates a list of worker nodes that
- * may need to access subplan results.
+ * may need to access subplan results. The function also sets writeToLocalFile
+ * flag if the result should also need be written locally.
  */
 List *
 FindAllWorkerNodesUsingSubplan(HTAB *intermediateResultsHash,
 							   char *resultId)
 {
-	List *workerNodeList = NIL;
 	IntermediateResultsHashEntry *entry =
 		SearchIntermediateResult(intermediateResultsHash, resultId);
 
+	List *remoteWorkerNodes = FindAllRemoteWorkerNodesUsingSubplan(entry);
+	if (remoteWorkerNodes == NIL)
+	{
+		/*
+		 * This could happen in two cases:
+		 * (a) Subquery in the having
+		 * (b) The intermediate result is not used, such as RETURNING of a
+		 *     modifying CTE is not used
+		 *
+		 * For SELECT, Postgres/Citus is clever enough to not execute the CTE
+		 * if it is not used at all, but for modifications we have to execute
+		 * the queries.
+		 */
+		entry->writeLocalFile = true;
+	}
+
+	/*
+	 * Don't include the current worker if the result will be written to local
+	 * file as this would be very inefficient and potentially leading race
+	 * conditions while tring to write the same file twice.
+	 */
+	if (entry->writeLocalFile)
+	{
+		remoteWorkerNodes = RemoveLocalNodeFromWorkerList(remoteWorkerNodes);
+	}
+
+	LogIntermediateResultMulticastSummary(entry, remoteWorkerNodes);
+
+	return remoteWorkerNodes;
+}
+
+
+/*
+ * FindAllRemoteWorkerNodesUsingSubplan goes over the nodeIdList of the
+ * intermediate result entry, and returns a list of workerNodes that the
+ * entry should be multi-casted to. The aim of the function is to filter
+ * out nodes with LOCAL_NODE_ID.
+ */
+static List *
+FindAllRemoteWorkerNodesUsingSubplan(IntermediateResultsHashEntry *entry)
+{
+	List *workerNodeList = NIL;
+
 	ListCell *nodeIdCell = NULL;
 	foreach(nodeIdCell, entry->nodeIdList)
 	{
-		WorkerNode *workerNode = LookupNodeByNodeId(lfirst_int(nodeIdCell));
+		uint32 nodeId = lfirst_int(nodeIdCell);
+		WorkerNode *workerNode = LookupNodeByNodeId(nodeId);
+		if (workerNode != NULL)
+		{
+			workerNodeList = lappend(workerNodeList, workerNode);
+		}
+	}
+
+	return workerNodeList;
+}
+
 
-		workerNodeList = lappend(workerNodeList, workerNode);
+/*
+ * RemoveLocalNodeFromWorkerList goes over the input workerNode list and
+ * removes the worker node with the local group id, and returns a new list.
+ */
+static List *
+RemoveLocalNodeFromWorkerList(List *workerNodeList)
+{
+	WorkerNode *workerNode = NULL;
+	int32 localGroupId = GetLocalGroupId();
 
-		if ((LogIntermediateResults && IsLoggableLevel(DEBUG1)) ||
-			IsLoggableLevel(DEBUG4))
+	/* we'll iterate over the list while deleting from it, so copy it */
```

Comment:
I don't like we need to copy the list and loop over it twice (`list_delete_ptr` loops again) just because we use `foreach_ptr`.

The postgres list api has functions to delete a cell while looping over it as can be seen here: https://github.com/postgres/postgres/blob/34f805c8cf1bc4d54075526d3b023d9194ccd2cd/contrib/file_fdw/file_fdw.c#L390-L409

```C
ListCell *workerNodeCell = NULL;
ListCell *prev = NULL;
foreach(workerNodeCell. workerNodeList)
{
	WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
	if (workerNode->groupId == localGroupId)
	{
		return list_delete_cell(workerNodeList, workerNodeCell, prev);
	}
	prev = workerNodeCell;
}
return workerNodeList;
```

should do the trick without copying and looping twice.

Reply:
done